### PR TITLE
feat(`forge`): new `flatten` implementation

### LINKED ISSUE #2344: Bug: Forge flatten does handle multiple differing SPDX licenses
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (3c49efe 2022-07-14T00:05:10.018576Z)

### What command(s) is the bug in?

forge build

### Operating System

macOS (Intel)

### Describe the bug

I run

```console
root:~ $ forge flatten src/3.sol > src/4.sol
```
contracts, 1, 2, 3, get flattened into 4.sol

Then when I run build I get this:

```console
root:~ $ forge build
[⠊] Compiling...
[⠒] Compiling 17 files with 0.8.15
[⠢] Solc 0.8.15 finished in 63.73ms
Error:
Compiler run failed
error[3716]: ParserError: Multiple SPDX license identifiers found in source file. Use "AND" or "OR" to combine multiple licenses. Please see https://spdx.org for more information.
```

Also I get this error message three times?

```console
warning: Implied profile [default] found in foundry.toml. This notation has been deprecated and may result in the profile not being registered in future versions. Please use [profile.default] instead.
warning: Implied profile [default] found in foundry.toml. This notation has been deprecated and may result in the profile not being registered in future versions. Please use [profile.default] instead.
warning: Implied profile [default] found in foundry.toml. This notation has been deprecated and may result in the profile not being registered in future versions. Please use [profile.default] instead.
```

### LINKED ISSUE #2378: Flatten removes "pragma experimental ABIEncoderV2;"
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (44f3a4b 2022-07-18T00:10:06.735136619Z)

### What command(s) is the bug in?

forge flatten

### Operating System

Linux

### Describe the bug

`forge flatten` will remove `pragma experimental ABIEncoderV2;` from the code and thus break it.

To reproduce I can run `cast etherscan-source -d etherscan/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0 0x7b2a3cf972c3193f26cdec6217d27379b6417bd0 ` to download https://etherscan.io/address/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0#code
The `ILendingPool.sol` will correctly contain `pragma experimental ABIEncoderV2;`.

If you now run:
```
forge flatten ./etherscan/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0/AToken/@aave/protocol-v2/contracts/protocol/tokenization/AToken.sol --output ./etherscan/0x7b2a3cf972c3193f26cdec6217d27379b6417bd0/Flattened.sol
```
The code will be flattened, but `pragma experimental ABIEncoderV2;` will be lost, breaking the code.

### LINKED ISSUE #2545: `forge flatten` does not support qualified imports
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (9ed1c37 2022-08-01T00:09:10.803586Z)

### What command(s) is the bug in?

forge flatten

### Operating System

macOS (Apple Silicon)

## Issue

When a flattened file contains a qualified import (of the form `import path as Q`), then:
1. `Q` is not stripped from accesses to definitions found in `path`. For instance, `Q.structName` does not become `structName`.
2. Names that clash with other names defined in other imported files are not modified to be unique. For instance, if the struct `DataType` is imported (with qualifier) from two files, compilation will fail.

## Suggestion
`forge flatten` should strip all qualifiers, and rename definitions as needed (recursively).

## Example:
```solidity
// contract.sol
pragma solidity ^0.8.13;
import "./aux1.sol" as A;
import "./aux2.sol" as B;

contract C {
  struct Data { A.DataType a; B.DataType b; }
}

// aux1.sol
struct DataType { string s; }

// aux2.sol
struct DataType { string s; }
```

`forge flatten` results in this (which fails to compile):

```solidity
pragma solidity ^0.8.13;
struct DataType { string s; }

struct DataType { string s; }

contract C {
  struct Data { A.DataType a; B.DataType b; }
}
```

when it should result in something like

```solidity
pragma solidity ^0.8.13;
struct DataType1 { string s; }

struct DataType2 { string s; }

contract C {
  struct Data { DataType1 a; DataType2 b; }
}
```

Note: this is kind of a followup to #1440.

### LINKED ISSUE #3265: `forge flatten` with renamed interface import does not compile
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (084ef3d 2022-09-19T00:04:35.98318Z)

### What command(s) is the bug in?

forge flatten

### Operating System

macOS (Apple Silicon)

### Describe the bug

If I have a contract using a named import of an interface, such as 
```solidity
import { IERC20 as MyIERC20 } from "src/interfaces/IERC20.sol";
```

`forge flatten` on that file results in output that includes the IERC20 interface, and the contract using the named import. Since the contract uses the named import `MyIERC20`, and the flattened file includes `interface IERC20` but not `interface MyIERC20`, `forge build` fails with `DeclarationError: Identifier not found or not unique.`

This seems related to https://github.com/foundry-rs/foundry/issues/1440

### LINKED ISSUE #3678: After run `forge flatten` command, then I run forge build, it shows: Multiple SPDX license identifiers found in source file.
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (c5791f5 2022-11-11T00:05:26.949466Z)

### What command(s) is the bug in?

forge flatten

### Operating System

macOS (Apple Silicon)

### Describe the bug

Firstly I run the `forge flatten --output src/a.sol  src/Impl/b.sol ` to flat my contract. 
Then I run `forge build`.
It will give me an error:  ParserError: Multiple SPDX license identifiers found in source file. Use "AND" or "OR" to combine multiple licenses.
截屏2022-11-13 上午11 21 23

### LINKED ISSUE #4034: forge flatten duplication of interfaces/contract names #262
### Component

Forge

### Have you ensured that all of these are up to date?

- [x] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (a44159a 2023-01-02T00:04:45.597843Z)

### What command(s) is the bug in?

forge flatten

### Operating System

macOS (Apple Silicon)

### Describe the bug

When running `forge flatten <file>` in a project that the path of files includes multiple interfaces/contracts with the same name, the command can't solve the duplication and just generate the output with the same name repeated.
For example if running `forge flatten` command in this file: https://github.com/makerdao/spells-goerli/blob/solc-0.8/src/DssSpell.sol will generate the result with two `OracleLike` interfaces. `Hevm` would have produced `OracleLike_1` and `OracleLike_2` instead.

### LINKED ISSUE #6022: Forge Flatten creates duplicate declarations (unable to compile)
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (95a93cd 2023-08-25T00:32:44.694202000Z)

### What command(s) is the bug in?

forge flatten

### Operating System

macOS (Apple Silicon)

### Describe the bug

To repro:

https://github.com/balancer/balancer-v2-monorepo/blob/master/pkg/vault/contracts/Vault.sol
Clone the repo and call `forge flatten Vault.sol  > VaultFlattened.sol`

Try to compile `VaultFlattened.sol` in a blank forge repo

Screen Shot 2023-10-10 at 5 22 40 PM

### LINKED ISSUE #6539: Sort before flatten
### Component

Forge

### Describe the feature you would like

`forge flatten` should sort interfaces/contracts/lib by name before flattening.

I've had a brief look at how this could be done, but it's non obvious to me:
```
/// Flattens all file imports into a single string
    pub fn flatten(&self, target: &Path) -> Result<String> {
        tracing::trace!("flattening file");
        let mut input_files = self.input_files();

        // we need to ensure that the target is part of the input set, otherwise it's not
        // part of the graph if it's not imported by any input file
        let flatten_target = target.to_path_buf();
        if !input_files.contains(&flatten_target) {
            input_files.push(flatten_target);
        }

        let sources = Source::read_all_files(input_files)?;
        let graph = Graph::resolve_sources(self, sources)?;
        self.flatten_node(target, &graph, &mut Default::default(), false, false, false).map(|x| {
            format!("{}\n", utils::RE_THREE_OR_MORE_NEWLINES.replace_all(&x, "\n\n").trim())
        })
    }
``` 
I don't know rust - potentially it could be as easy as sorting nodes before transforming to a string 

### Additional context

I'm a heavy user of `forge flatten` as it's quite nice to compare an upstream deployment of a contract with a local version.
Diffing folders usually is a lot harder then diffing single files.

That said, diffing flattened files can become problematic, when - for whatever reason - the import order on one of the two sides changes. The diff will grow huge and unreadable. If the contracts and interfaces would be sorted alphabetically on the flattened files it would be far easier to understand differences.

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Ref https://github.com/foundry-rs/compilers/pull/52
Full rewrite of flattening utilizing native solc AST.

- Closes #6539 
- Fixes #6022 
- Fixes #4034 
- Fixes #3678 
- Fixes #3265 
- Fixes #2545 
- Fixes #2378 
- Fixes #2344

## Solution

It seems that we don't have any tests for `forge flatten` but creating PR with patched dep to run CI just in case.

### CODE REVIEW

**Comment by DaniPopes** [comment]
(You have to write closes before each PR, preferably each on a new line)

**Comment by avniculae** [comment]
@klkvr Great work, congrats! I run your version of flatten on a project that uses prb math (both UD60x18 and SD59x18) and I can see that the resulted file has some issues renamed variables in assembly blocks. For example, in the code below uUNIT was not replaced with the renamed version (uUNIT_1, _2, ...)
```
assembly ("memory-safe") {
        switch x
        case 1 { result := mul(uUNIT, sub(0, 18)) }
        case 10 { result := mul(uUNIT, sub(1, 18)) }
        case 100 { result := mul(uUNIT, sub(2, 18)) }
        case 1000 { result := mul(uUNIT, sub(3, 18)) }
        case 10000 { result := mul(uUNIT, sub(4, 18)) }
.....
```

Is this an easy fix for you?

**Review by mattsse — APPROVED** [review]
lgtm!!

nice work

File: crates/forge/bin/cmd/flatten.rs [hunk 1]
```diff
@@ -38,10 +39,27 @@ impl FlattenArgs {
...
+                Ok(Flattener::new(&project, &compiler_output, &target_path)?.flatten())
+            }
+            Err(_) => {
+                // Fallback to the old flattening compilation if we couldn't compile the target
+                // successfully. This would be the case if the target has invalid
+                // syntax. (e.g. Solang)
+                Ok(project.paths.flatten(&target_path)?)
+            }
+        }
+        .map_err(|err: SolcError| eyre::eyre!("Failed to flatten: {err}"))?;
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/flatten.rs:62]: this map_err is never reached because of the `?`

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Comment by avniculae** [comment]
> @avniculae ah, assembly blocks are not in scope of renaming atm
> 
> It shouldn't be hard to fix, could you please drop a link to the contract you are flattening or a minimal repro so it's easier to debug?

Thanks for letting me know! I run on a private repo, but I think [this file](https://github.com/sablier-labs/v2-core/blob/main/src/SablierV2LockupDynamic.sol) should represent a good test.

**Comment by avniculae** [comment]
@klkvr another point: if the contract renames an import (e.g., `import { mul as mulUD } from "..."`), the flatten command fails to rename `mulUD` to the new renaming

**Comment by klkvr** [comment]
@avniculae thanks for all the points, will try to fix all of them

regarding imports renaming could you please give a bit more context as this case was addressed directly during development and it shouldn't fail

which entity is the mulUD? I mean variable/enum/struct etc

**Comment by avniculae** [comment]
@klkvr I am experiencing issues with function renaming (mul is a function here)

**Comment by klkvr** [comment]
@avniculae 

I can't reproduce this behavior locally, maybe I got something wrong?
So, I am creating 2 files:

mul.sol:
```solidity
function mul(uint256 a, uint256 b) pure returns(uint256) {
    return a * b;
}
```
test.sol
```solidity
pragma solidity 0.8.6;

import {mul as mulUD} from "./mul.sol";

contract Test {
    function foo() public {
        mulMD(1, 2);
    }
}
```

And test.sol is getting flattened to 
```solidity
pragma solidity 0.8.6;
function mul(uint256 a, uint256 b) pure returns(uint256) {
    return a * b;
}

uint256 constant someVar = 1;

contract Test {
    function foo(uint256 a) public {
        mul(1, 2);
    }
}
```

which seems correect

File: crates/forge/bin/cmd/flatten.rs [hunk 2]
```diff
@@ -38,10 +39,28 @@ impl FlattenArgs {
...
+        let sources = Source::read_all(vec![target_path.clone()])?;
+        let (sources, _) = Graph::resolve_sources(&project.paths, sources)?.into_sources();
+
+        let compiler_output = ProjectCompiler::new().files(sources.into_keys()).compile(&project);
+
+        let flattened = match compiler_output {
+            Ok(compiler_output) => match Flattener::new(&project, &compiler_output, &target_path) {
+                Ok(flattener) => Ok(flattener.flatten()),
+                Err(err) => Err(err),
+            },
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/flatten.rs:55]: `new().map()`?

File: crates/forge/bin/cmd/flatten.rs [hunk 3]
```diff
@@ -38,10 +39,28 @@ impl FlattenArgs {
 
         let config = build_args.try_load_config_emit_warnings()?;
 
-        let paths = config.project_paths();
         let target_path = dunce::canonicalize(target_path)?;
-        let flattened =
-            paths.flatten(&target_path).map_err(|err| eyre::eyre!("Failed to flatten: {err}"))?;
+
+        // We need to provide Flattener with compiled output of target and all of it's imports.
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/flatten.rs:44]:
```suggestion
        // We need to provide Flattener with compiled output of target and all of its imports.
```

**Comment by avniculae** [comment]
@klkvr Here's a small counterexample:

**SFixed.sol**
```
pragma solidity 0.8.19;

type SFixed256x18 is int256;

using {add} for SFixed256x18 global;

function add(SFixed256x18 a, SFixed256x18 b) pure returns (SFixed256x18) {
    return SFixed256x18.wrap(SFixed256x18.unwrap(a) + SFixed256x18.unwrap(b));
}
```

**UFixed.sol**
```
pragma solidity 0.8.19;

type UFixed256x18 is uint256;

using {add} for UFixed256x18 global;

function add(UFixed256x18 a, UFixed256x18 b) pure returns (UFixed256x18) {
    return UFixed256x18.wrap(UFixed256x18.unwrap(a) + UFixed256x18.unwrap(b));
}
```

**Contract.sol**
```
pragma solidity 0.8.19;

import {SFixed256x18} from "./SFixed.sol";
import {UFixed256x18} from "./UFixed.sol";

contract Test {
    function fun() public pure returns (SFixed256x18, UFixed256x18) {
      SFixed256x18 a = SFixed256x18.wrap(10);
      SFixed256x18 b = SFixed256x18.wrap(-90);
      SFixed256x18 c = a.add(b);

      UFixed256x18 d = UFixed256x18.wrap(20);
      UFixed256x18 e = UFixed256x18.wrap(30);
      UFixed256x18 f = d.add(e);

      return (c, f);
    }
}
```

In the flattened version, `a.add(b)` is replaced by `add_0(b)`, while `d.add(e)` is replaced by `add_1(e)`. However, on the project I'm running it (which uses PRB-Math heavily), the command often leaves these functions unchanged ( e.g., `a.add(b)`).

**Review by mattsse — APPROVED** [review]
🔥

**Comment by klkvr** [comment]
hey @avniculae thank you once again for all the edge cases

I belive that all of them should be fixed now, could you please pull this PRs version and try once again if they are working for you?

File: crates/forge/bin/cmd/flatten.rs [hunk 4]
```diff
@@ -38,10 +39,27 @@ impl FlattenArgs {
...
         let target_path = dunce::canonicalize(target_path)?;
-        let flattened =
-            paths.flatten(&target_path).map_err(|err| eyre::eyre!("Failed to flatten: {err}"))?;
+
+        // We need to provide Flattener with compiled output of target and all of its imports.
+        let project = config.ephemeral_no_artifacts_project()?;
+        let sources = Source::read_all(vec![target_path.clone()])?;
+        let (sources, _) = Graph::resolve_sources(&project.paths, sources)?.into_sources();
+
+        let compiler_output = ProjectCompiler::new().files(sources.into_keys()).compile(&project);
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/flatten.rs:49]: can't we just use `compile` without doing the Source and Graph stuff?

Comment (klkvr) [reply crates/forge/bin/cmd/flatten.rs]: yeah, it seems that we can. I belive that was a workaround I used before #6935 which is no more needed

**Comment by avniculae** [comment]
@klkvr Thanks for pushing these fixes! I re-run the latest version and I now get the following error:
```
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
Location: /Users/aniculae/.cargo/registry/src/index.crates.io-6f17d22bba15001f/foundry-compilers-0.3.0/src/flatten.rs:450

This is a bug. Consider reporting it at https://github.com/foundry-rs/foundry

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 8 frames hidden ⋮                               
   9: core::panicking::panic::h40561ff494e2b577
      at <unknown source file>:<unknown line>
  10: foundry_compilers::flatten::Flattener::flatten::hf4ff1bfccf30ce3e
      at <unknown source file>:<unknown line>
  11: forge::main::h82dc03aed91a2c0a
      at <unknown source file>:<unknown line>
  12: std::sys_common::backtrace::__rust_begin_short_backtrace::hb40b21b17ab48be9
      at <unknown source file>:<unknown line>
  13: std::rt::lang_start::h39e6cad84a7b22b7
      at <unknown source file>:<unknown line>
  14: _main<unknown>
      at <unknown source file>:<unknown line>
```
However, it's hard to tell what's causing this error since the flattened code has over 11k loc, and I wasn't able to reproduce it with small examples

**Comment by avniculae** [comment]
@klkvr Thanks for pushing these fixes! I re-run the latest version and I now get the following error:
```
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
Location: .../foundry-compilers-0.3.0/src/flatten.rs:450

This is a bug. Consider reporting it at https://github.com/foundry-rs/foundry

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 8 frames hidden ⋮                               
   9: core::panicking::panic::h40561ff494e2b577
      at <unknown source file>:<unknown line>
  10: foundry_compilers::flatten::Flattener::flatten::hf4ff1bfccf30ce3e
      at <unknown source file>:<unknown line>
  11: forge::main::h82dc03aed91a2c0a
      at <unknown source file>:<unknown line>
  12: std::sys_common::backtrace::__rust_begin_short_backtrace::hb40b21b17ab48be9
      at <unknown source file>:<unknown line>
  13: std::rt::lang_start::h39e6cad84a7b22b7
      at <unknown source file>:<unknown line>
  14: _main<unknown>
      at <unknown source file>:<unknown line>
```
However, it's hard to tell what's causing this error since the flattened code has over 11k loc, and I wasn't able to reproduce it with small examples

**Comment by mattsse** [comment]
@avniculae is this repo public?

**Comment by klkvr** [comment]
@avniculae this happens while updating `@inheritdoc` tags, do you have any uncommon patters in those? Like using aliases, qualified imports? Or all values are just straight contract/interface names?

**Comment by avniculae** [comment]
@klkvr They should all be standard, like this example (I'm less sure about dependencies though)
```
  /**
   * @inheritdoc InterfaceName
   */
```

**Comment by klkvr** [comment]
@avniculae any chance you can share names of other dependencies besides prb-math?

**Comment by avniculae** [comment]
@klkvr I think the main ones are prb-math and @openzeppelin/contracts/

**Review by mattsse — APPROVED** [review]

**Comment by mattsse** [comment]
This includes a ton of improvements, but might still have some limitations which we can fix as soon as they get reported.

**Pull request merged**